testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_transactions.py

```diff
@@ -0,0 +1,420 @@
+    cb_server = couchbaseserver.CouchbaseServer(cbs_one_url)
+
+    sg_db = "db"
+    username = "autotest"
+    password = "password"
+    channels = ["TRANSACTIONS"]
+    num_of_docs = 4
+    doc_id = "transaction-sync-id"
+    doc_ids = ""
+
+    sg_client = MobileRestClient()
+    doc_obj = Document(base_url)
+    # Reset cluster to ensure no data in system
+    c = cluster.Cluster(config=cluster_config)
+    c.reset(sg_config_path=sg_config)
+    bucket = cb_server.get_bucket_names()[0]
+
+    # disable replicas and rebalance the server
+    cb_server.disable_replicas(bucket)
+    cb_server.rebalance_server(cluster_servers)
+
+    # randomly  generate replication ids
+    for i in range(0, num_of_docs):
+        doc_id = "txn_id_" + str(time.time())
+        doc_ids = doc_id + "," + doc_ids
+    # 1. Insert a transaction with 3 or more docs
+    opt_doc_ids = doc_ids.replace("txn_id_", "insert-txn_id_")
+    transactions_app_dir = "resources/data/transaction-performer.jar"
+    doc_body = document.create_doc(doc_id=doc_id, content="doc1", channels=channels)
+    data = json.dumps(doc_body)
+    cmd = "java -cp {} com.couchbase.transaction clusterip={} operationdocids={} bucket={} doccontent='{}'".format(transactions_app_dir, c.servers[0].host, opt_doc_ids, bucket, data)
+    subprocess.check_output(cmd, shell=True)
+
+    # 2. replication from server to SGW and to CBL"
+    # 3. Verify all docs in transactions are pulled to CBL
+    replicator = Replication(base_url)
+    sg_client.create_user(sg_admin_url, sg_db, username, password, channels=channels)
+    _, _, repl = replicator.create_session_configure_replicate(
+        base_url, sg_admin_url, sg_db, username, password, channels, sg_client, cbl_db, sg_blip_url, continuous=True, replication_type="push_pull")
+    replicator.wait_until_replicator_idle(repl)
+    orig_cbl_doc_ids = db.getDocIds(cbl_db)
+    assert len(orig_cbl_doc_ids) == num_of_docs, "docs from committed transaction did not replicated to cbl"
+
+    # 4. Replace the transaction with new content
+    list_replace_ids = random.sample(orig_cbl_doc_ids, 2)
+    replace_ids = ""
+    for rep_id in list_replace_ids:
+        replace_ids = rep_id + "," + replace_ids
+    opt_doc_ids = replace_ids.replace("txn_id_", "replace-txn_id_")
+    cmd = "java -cp {} com.couchbase.transaction clusterip={} operationdocids={} bucket={}".format(transactions_app_dir, c.servers[0].host, opt_doc_ids, bucket)
+    subprocess.check_output(cmd, shell=True)
+    replicator.wait_until_replicator_idle(repl)
+    cbl_doc_ids = db.getDocIds(cbl_db)
+    for id in list_replace_ids:
+        cbl_doc = db.getDocument(cbl_db, id)
+        doc_mut = doc_obj.toMutable(cbl_doc)
+        doc_body1 = doc_obj.toMap(doc_mut)
+        assert("_updated" in doc_body1["id"])
+        assert("_updated" in doc_body1["content"])
+
+    # 6. Delete the transaction
+    # 7. Verify doc is removed in cbl
+    list_deleted_ids = random.sample(orig_cbl_doc_ids, 2)
```

Comment:
is that possible that the ids are selected for deletion have been updated in previous replace step, and you are no longer able to find in the current db, therefore looks deletion is done but actually not found but nothing to delete?
does orig_cbl_doc_ids need to be updated after step 4/5?

Reply:
no, when doc updated in previous replace step, it updates the content, but not doc id. So, when you try to delete from transaction app, u delete by doc id(same concept as sgw and cbl) , but not by doc content. After you delete the doc from transaction app side, replication continues to cbl and test checks whether doc id still exists in cbl. If doc id still exists, transaction app failed to remove. Expectation is if doc id removed in transaction app, it should get removed in cbl after replication.  If you still have question on this, I can explain next week